server/engine/dryrun.go

```diff
@@ -0,0 +1,210 @@
+	"github.com/microsoft/commercial-marketplace-offer-deploy/sdk"
+	log "github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+var (
+	dryRunInstance     *dryRunController
+	dryRunInstanceOnce sync.Once
+	dryRunInstanceErr  error
+)
+
+type ErrorHandler func(err error)
+type NameOrKeyCreate func() string
+
+type dryRunController struct {
+	// the MODM deployment id
+	deploymentId   int
+	db             *gorm.DB
+	done           chan struct{}
+	clientEndpoint string
+	location       string
+	resourceGroup  string
+	subscription   string
+	apiKey         string
+	hookName       string
+	deploymentName string
+
+	// this is the url that will be called by MODM. It maps to /eventhook route for handler/eventhook
+	eventHookCallbackUrl string
+	HandleError          ErrorHandler
+}
+
+func (d *dryRunController) Execute(ctx context.Context) {
+	time.Sleep(10 * time.Second)
+
+	go func() {
+		step, err := d.getStep()
+		if err != nil {
+			d.HandleError(err)
+		}
+
+		cred, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			d.HandleError(err)
+		}
+
+		client, err := sdk.NewClient(d.clientEndpoint, cred, nil)
+		if err != nil {
+			d.HandleError(err)
+		}
+
+		deploymentName := d.deploymentName
+		request := api.CreateDeployment{
+			Name:           &deploymentName,
+			Template:       step.Template,
+			Location:       &d.location,
+			ResourceGroup:  &d.resourceGroup,
+			SubscriptionID: &d.subscription,
+		}
+
+		dep, err := client.Create(ctx, request)
+		if err != nil {
+			d.HandleError(err)
+		}
+		d.deploymentId = int(*dep.ID)
+
+		createEventRequest := api.CreateEventHookRequest{
+			APIKey:   &d.apiKey,
+			Callback: &d.eventHookCallbackUrl,
+			Name:     &d.hookName,
+		}
+
+		_, err = client.CreateEventHook(ctx, createEventRequest)
+		if err != nil {
+			d.HandleError(err)
+		}
+
+		res, err := client.DryRun(ctx, d.deploymentId, step.Parameters)
+		if err != nil {
+			d.HandleError(err)
+		}
+
+		d.create(res)
+	}()
+
+	<-d.done
+}
+
+func DryRunControllerInstance() (*dryRunController, error) {
+	dryRunInstanceOnce.Do(func() {
+		dryRunInstance = &dryRunController{
+			db:                   persistence.NewPersistentDB(config.GetEnvironment().DB_PATH).Instance,
+			resourceGroup:        config.GetEnvironment().RESOURCE_GROUP_NAME,
+			subscription:         config.GetEnvironment().SUBSCRIPTION,
+			location:             config.GetEnvironment().AZURE_LOCATION,
+			apiKey:               config.GetEnvironment().WEB_HOOK_API_KEY,
+			hookName:             "deployment-driver-hook",
+			deploymentName:       "deployment-driver/" + uuid.New().String(),
+			eventHookCallbackUrl: config.GetEnvironment().WEB_HOOK_CALLBACK_URL,
+			clientEndpoint:       "http://localhost:8080",
+			done:                 make(chan struct{}),
+			HandleError: func(err error) {
+				if err != nil {
+					log.Error(err)
+				}
+			},
+		}
+	})
+	return dryRunInstance, dryRunInstanceErr
+}
+
+func DryRunDone(eventHook *events.EventHookMessage) {
```

Comment:
This wrapper is just returns the channel for controller being done, we might as well export the `dryRunController.dryRunDone()` itself and get rid of this wrapper.